### PR TITLE
Misc changes related to Miri allocations

### PR DESCRIPTION
**Pull request by Zoxc**
This builds on top of https://github.com/rust-lang/rust/pull/50249

r? @oli-obk

### CODE REVIEW

**Comment by pietroalbini** [comment]
Ping from triage @oli-obk! This PR needs your review.
@Zoxc there are some merge conflicts.

**Review by oli-obk — CHANGES_REQUESTED** [review]
As discussed on IRC I am very unsure about the last commit. We had a scheme like that when miri was merged, and we kept running into various edge cases with the decoding order. Even creating MCVEs for the panics we were getting was hard, because small changes in the code would change the order of evaluation.

cc @michaelwoerister

File: src/librustc/ich/impls_ty.rs [hunk 1]
```diff
@@ -439,30 +428,29 @@ impl<'a> HashStable<StableHashingContext<'a>> for mir::interpret::AllocId {
...
-            } else {
-                bug!("no allocation for {}", self);
-            }
+            let alloc_kind = tcx.alloc_map.lock().get(*self).expect("no value for AllocId");
+            alloc_kind.hash_stable(hcx, hasher);
         });
     }
 }
 
+impl<'a, 'gcx, M: HashStable<StableHashingContext<'a>>> HashStable<StableHashingContext<'a>>
```
[the remark this answers is not in the document]
Comment (Zoxc) [reply src/librustc/ich/impls_ty.rs]: I don't think we have a macro which works for enums with lifetimes.

**Comment by Zoxc** [comment]
I've split the last commit into a separate PR.

File: src/bootstrap/lib.rs [hunk 2]
```diff
@@ -592,12 +592,20 @@ impl Build {
             Path::new(llvm_bindir.trim()).join(exe("FileCheck", &*target))
         } else {
             let base = self.llvm_out(self.config.build).join("build");
-            let exe = exe("FileCheck", &*target);
-            if !self.config.ninja && self.config.build.contains("msvc") {
```
Comment (oli-obk) [inline src/bootstrap/lib.rs]: Please put this in a separate PR, I have no idea what this does or why it's here

**Review by oli-obk — APPROVED** [review]

**Comment by alexcrichton** [comment]
FWIW this hurt the max-rss usage on a [few benchmarks](https://perf.rust-lang.org/compare.html?start=1bbae5f38677e823ba6e23f1e0e105ceee4c6f8a&end=ff8fa5cc69db5567b32ceca1ee4ac0dcfa3a81bc&stat=max-rss) (but not by huge amounts)

**Pull request merged**